components/spinner.py

```diff
@@ -3,29 +3,128 @@
+    ROTATION_SPIN_FACTOR = 20  # increase this to recude slow down
+    ROTATION_MAX_SPEED = 1
+    ROTATION_TARGET_ROTATIONS = 3.5
+
     def __init__(self):
         pass
 
     def setup(self):
-        pass
+        self.spinner_motor.set(0)
+        self.raise_wheel()
+        self.rotations = 0
+        # self.state = None
+        self.piston_state = "down"
+        self.required_colour = "B"
+        self.lastCol = "Y"
 
-    def on_enable(self):
-        self.spinner_motor.stopMotor()
-        self.spinner_solenoid.set(wpilib.DoubleSolenoid.Value.kReverse)
+    def on_enable(self) -> None:
+        self.spinner_motor.set(0)
+        self.raise_wheel()
 
-    def execute(self):
-        pass
+    def raise_wheel(self) -> None:  # moves wheel up
+        self.spinner_solenoid.set(False)
+        self.piston_state = "up"
 
-    def read_colour(self):
-        return (self.colour_sensor.getColor(), self.colour_sensor.getIR())
+    def lower_wheel(self) -> None:  # moves wheel down
+        self.spinner_solenoid.set(True)
+        self.piston_state = "down"
 
-    def piston_up(self):
-        self.spinner_solenoid.set(wpilib.DoubleSolenoid.Value.kForward)
+    def get_current_colour(self) -> str:  # reads and catagorizes colour
+        self.sensed_colour = self.colour_sensor.getColor()
 
-    def piston_down(self):
-        self.spinner_solenoid.set(wpilib.DoubleSolenoid.Value.kReverse)
+        distances = {}
+        for col_name, col_value in self.WHEEL_COLOURS.items():
+            distances[col_name] = col_value.dist(self.sensed_colour)
+
+        if (
+            distances[min(distances, key=lambda col: distances[col])]
+            < self.MAX_COLOUR_DIST
+        ):
+            return min(distances, key=lambda col: distances[col])
+
+    def get_wheel_dist(self) -> int:
+        # find the distance of the colour wheel from desired state in segments
+        # reuturns between -2 and 2 with 0 being the colour desired
+        current_colour = (
+            self.get_current_colour()
+        )  # only called when running position control
+        if current_colour == None:
+            return 0
+        distance = (
+            self.WHEEL_ORDER.index(self.required_colour)
+            + self.COLOUR_OFFSET
+            - self.WHEEL_ORDER.index(current_colour)
+        )
+        if distance > 2:
+            return distance - 4
+        return distance
+
+    def go_to_colour(self, colour: str) -> bool:  # do position control
+        self.required_colour = colour
+        distance = self.get_wheel_dist()
+        self.spinner_motor.set(distance * self.POSITION_SPIN_FACTOR)
+
+        if distance == 0:
+            return True
+        else:
+            return False
+
+    def do_rotation(self) -> bool:
+        spin_speed = 1 - (
+            self.ROTATION_SPIN_FACTOR
+            ** (self.rotations - self.ROTATION_TARGET_ROTATIONS)
+            * self.ROTATION_MAX_SPEED
+        )
```

Comment:
This does not currently ensure that the speed is below 60rpm, if you take the diameters of your wheel as constants you can use your assume the your worst case displacement on the wheel to do this calculation